File: src/low_level/network.jl
export
		add_arc,
		add_node,
		get_all_nodes,
		get_children,
		get_first_node,
		get_next_node,
		get_node,
		get_number_of_nodes,
		get_parents,
		find_node,
		is_acyclic,
		read_file,
		write_file,
		set_default_BN_algorithm,
		set_default_ID_algorithm,
		set_default_HBN_algorithm,
		get_default_BN_algorithm,
		get_default_ID_algorithm,
		get_default_HBN_algorithm,
		DSL_ALG_BN_LAURITZEN, DSL_ALG_BN_HENRION, DSL_ALG_BN_PEARL, DSL_ALG_BN_LSAMPLING,
		DSL_ALG_BN_SELFIMPORTANCE, DSL_ALG_BN_HEURISTICIMPORTANCE, DSL_ALG_BN_BACKSAMPLING,
		DSL_ALG_BN_AISSAMPLING, DSL_ALG_BN_EPISSAMPLING, DSL_ALG_BN_LBP, DSL_ALG_BN_LAURITZEN_OLD,
		DSL_ALG_BN_RELEVANCEDECOMP, DSL_ALG_BN_RELEVANCEDECOMP2, DSL_ALG_HBN_HEPI,
		DSL_ALG_HBN_HLW, DSL_ALG_HBN_HLBP, DSL_ALG_HBN_HLOGICSAMPLING,
		update_beliefs,
		invalidate_all_beliefs,
		call_ID_algorithm,
		call_BN_algorithm,
		call_Eq_algorithm,
		call_HBN_algorithm,
		clear_all_evidence,
		clear_all_decision,
		clear_all_propagated_evidence,
		is_there_any_evidence,
		is_there_any_decision,
		partial_ordering

function add_arc( net::Network, theParent::Integer, theChild::Integer )
	err_int = ccall( (:network_AddArc, LIB_SMILE), Int32, (Ptr{Void}, Int32, Int32), net.ptr, theParent, theChild)
	if err_int < 0
		error( "network_AddArc: code: ", err_int )
	end
end

function add_node( net::Network, nodeType::Integer, nodeName::AbstractString )

	if !is_nodetype_id(nodeType)
		error( "network_AddNode: nodeType not a recognized type" )
	end

	error_buffer = Array(UInt8, 128)
	nodeId = ccall( (:network_AddNode, LIB_SMILE), Int32, (Ptr{Void}, Int32, Ptr{UInt8}, Ptr{UInt8}), net.ptr, nodeType, bytestring(nodeName), error_buffer )
	if nodeId < 0
		error_str = bytestring(error_buffer)
		error( "network_AddNode: ", error_str )
	end
	return nodeId
end

function get_all_nodes( net::Network )

	n = get_number_of_nodes(net)
	if n == 0
		return Int[]
	end

	nodeHandles = Array(Int,n)
	nodeHandles[1] = get_first_node( net )
	for i = 2 : n
		nodeHandles[i] = get_next_node( net, nodeHandles[i-1] )
	end

	return nodeHandles
end

function get_children( net::Network, ofThisNode::Integer, len_start = 10 )
	len = Array(UInt32,1)
	len[1] = len_start
	arr = Array(Int32, len_start)
	ccall( (:network_GetChildren, LIB_SMILE), Void, (Ptr{Void},Int32,Ptr{Int32},Ptr{UInt32}), net.ptr, ofThisNode, arr, len )
	len_end = len[1]

	if len_end == 0
		return []
	end
	if len_end > len_start
		error("get_children:: not enough spaces allocated in array")
	end
	return arr[1:len_end]
end

function get_first_node( net::Network )
	retval = ccall( (:network_GetFirstNode, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
	return retval
end

function get_next_node( net::Network, ofThisNode::Integer )
	retval = ccall( (:network_GetNextNode, LIB_SMILE), Int32, (Ptr{Void},Int32), net.ptr, ofThisNode )
	return retval
end

function get_node( net::Network, nodeHandle::Integer )
	ptr = ccall( (:network_GetNode, LIB_SMILE), Ptr{Void}, (Ptr{Void},Int32), net.ptr, nodeHandle )
	return Node(ptr)
end

function get_number_of_nodes( net::Network )
	retval = ccall( (:network_GetNumberOfNodes, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
	return retval
end

function get_parents( net::Network, ofThisNode::Integer, len_start = 10 )
	len = Array(UInt32,1)
	len[1] = len_start
	arr = Array(Int32, len_start)
	ccall( (:network_GetParents, LIB_SMILE), Void, (Ptr{Void},Int32,Ptr{Int32},Ptr{UInt32}), net.ptr, ofThisNode, arr, len )
	len_end = len[1]

	if len_end == 0
		return []
	end
	if len_end > len_start
		error("get_parents:: not enough spaces allocated in array")
	end
	return arr[1:len_end]
end

find_node( net::Network, withThisID::AbstractString ) =
	ccall( (:network_FindNode, LIB_SMILE), Int32, (Ptr{Void},Ptr{UInt8}), net.ptr, bytestring(withThisID) )

function is_acyclic( net::Network )
	retval = ccall( (:network_IsAcyclic, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
	return retval != 0 # true if non zero
end

function read_file( net::Network, fileName::AbstractString )
	# TODO: check that fileName ends with a valid file type
	retval = ccall( (:network_ReadFile, LIB_SMILE), Int32, (Ptr{Void},Ptr{UInt8}), net.ptr, bytestring(fileName) )
	return retval
end

function write_file( net::Network, fileName::AbstractString )
	ext = splitext(fileName)[2]
	@assert(ext == ".dsl" || ext == ".xdsl")

	retval = ccall( (:network_WriteFile, LIB_SMILE), Int32, (Ptr{Void},Ptr{UInt8}), net.ptr, bytestring(fileName) )
	return retval
end

# BN algorithms
const DSL_ALG_BN_LAURITZEN            = Int32( 0)
const DSL_ALG_BN_HENRION              = Int32( 1)
const DSL_ALG_BN_PEARL                = Int32( 2)
const DSL_ALG_BN_LSAMPLING            = Int32( 3)
const DSL_ALG_BN_SELFIMPORTANCE       = Int32( 4)
const DSL_ALG_BN_HEURISTICIMPORTANCE  = Int32( 5)
const DSL_ALG_BN_BACKSAMPLING         = Int32( 6)
const DSL_ALG_BN_AISSAMPLING          = Int32( 7)
const DSL_ALG_BN_EPISSAMPLING         = Int32( 8)
const DSL_ALG_BN_LBP				          = Int32( 9)
const DSL_ALG_BN_LAURITZEN_OLD        = Int32(10)
const DSL_ALG_BN_RELEVANCEDECOMP      = Int32(11)
const DSL_ALG_BN_RELEVANCEDECOMP2     = Int32(12)
const DSL_ALG_HBN_HEPIS				        = Int32(13)
const DSL_ALG_HBN_HLW                 = Int32(14)
const DSL_ALG_HBN_HLBP				        = Int32(15)
const DSL_ALG_HBN_HLOGICSAMPLING	    = Int32(16)

# ID algorithms
const DSL_ALG_ID_COOPERSOLVING        = Int32(0)
const DSL_ALG_ID_SHACHTER             = Int32(1)

set_default_BN_algorithm(net::Network,  algorithm::Integer) = ccall( (:network_SetDefaultBNAlgorithm,  LIB_SMILE), Void, (Ptr{Void},Int32), net.ptr, algorithm )
set_default_ID_algorithm(net::Network,  algorithm::Integer) = ccall( (:network_SetDefaultIDAlgorithm,  LIB_SMILE), Void, (Ptr{Void},Int32), net.ptr, algorithm )
set_default_HBN_algorithm(net::Network, algorithm::Integer) = ccall( (:network_SetDefaultHBNAlgorithm, LIB_SMILE), Void, (Ptr{Void},Int32), net.ptr, algorithm )
get_default_BN_algorithm(net::Network)  = ccall( (:network_GetDefaultBNAlgorithm,  LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
get_default_ID_algorithm(net::Network)  = ccall( (:network_GetDefaultIDAlgorithm,  LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
get_default_HBN_algorithm(net::Network) = ccall( (:network_GetDefaultHBNAlgorithm, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )

# Inference
update_beliefs(net::Network)         = ccall( (:network_UpdateBeliefs,        LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
invalidate_all_beliefs(net::Network) = ccall( (:network_InvalidateAllBeliefs, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
call_ID_algorithm(net::Network)      = ccall( (:network_CallIDAlgorithm,      LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
call_BN_algorithm(net::Network)      = ccall( (:network_CallBNAlgorithm,      LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
call_Eq_algorithm(net::Network)      = ccall( (:network_CallEqAlgorithm,      LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
call_HBN_algorithm(net::Network)     = ccall( (:network_CallHBNAlgorithm,     LIB_SMILE), Int32, (Ptr{Void},), net.ptr )

clear_all_evidence(net::Network)            = ccall( (:network_ClearAllEvidence, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
clear_all_decision(net::Network)            = ccall( (:network_ClearAllDecision, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
clear_all_propagated_evidence(net::Network) = ccall( (:network_ClearAllPropagatedEvidence, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
is_there_any_evidence(net::Network)         = ccall( (:network_IsThereAnyEvidence, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )
is_there_any_decision(net::Network)         = ccall( (:network_IsThereAnyDecision, LIB_SMILE), Int32, (Ptr{Void},), net.ptr )

partial_ordering(net::Network) = IntArray(ccall( (:network_PartialOrdering, LIB_SMILE), Ptr{Void}, (Ptr{Void},), net.ptr ))

File: src/wrapper/cpp/network.cpp
// SMILE.JL
// A C library wrapper for SMILE (Structural Modeling, Inference, and Learning Engine)
// so that users may call SMILE functions through Julia
//
// author: Tim Wheeler
// date: October 2014
// Stanford Intelligence Systems Laboratory

// #include <stdio.h>

///////////////////////////////////////
//            DSL_NETWORK            //
///////////////////////////////////////

void * createNetwork()
{
	DSL_network * theNet = new DSL_network();
	void * retval = theNet;
	return retval;
}

void freeNetwork(void * void_net)
{
    DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	delete net;
}

int network_AddArc(void * void_net, int theParent, int theChild)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->AddArc( theParent, theChild );
}

int network_AddNode(void * void_net, int thisType, char *thisId, char *exception)
{
	/*
	DSL_network.AddNode( int thisType, char* thisId)
	Adds a node of thisType to the network. thisId indicates the identifier of the new node.
	If thisId is NULL, a unique identifier will be generated.
	If the node is correctly created, the returning value will be the handle of the node
	This handle is guaranteed to remain constant for the lifetime of the network

	Error Values:
	DSL_OUT_OF_MEMORY
	*/
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	int uid = -1;
	uid = net->AddNode( thisType, thisId );
	
	if (uid < 0)
	{
		if ( uid == DSL_OUT_OF_MEMORY )
			sprintf( exception, "DSL_OUT_OF_MEMORY" );
		else if ( uid == DSL_OUT_OF_RANGE )
			sprintf( exception, "DSL_OUT_OF_RANGE" ); // thisId is already in use
		else
			sprintf( exception, "UNKNOWN_ERROR" );
	}

	return uid;
}

void network_GetChildren(void * void_net, int ofThisNode, int * childHandles, unsigned  * len)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	DSL_intArray children = net->GetChildren(ofThisNode);
	for ( unsigned i = 0; i < (unsigned)(children.NumItems()) && i < *len; i++ )
	{
		childHandles[i] = children[i];
	}
	*len = (unsigned)(children.NumItems());
}

int network_GetFirstNode( void * void_net )
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->GetFirstNode();
}

void network_GetParents(void * void_net, int ofThisNode, int * parentHandles, unsigned  * len)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	DSL_intArray parents = net->GetParents(ofThisNode);
	for ( unsigned i = 0; i < (unsigned)(parents.NumItems()) && i < *len; i++ )
	{
		parentHandles[i] = parents[i];
	}
	*len = (unsigned)(parents.NumItems());
}

int network_FindNode(void * void_net, char *withThisId)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->FindNode(withThisId);
}

int network_GetNextNode( void * void_net, int ofThisNode )
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->GetNextNode(ofThisNode);
}

void * network_GetNode(void * void_net, int theNode)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	DSL_node * node = net->GetNode(theNode);
	void * retval = node;
	return retval;
}

int network_GetNumberOfNodes( void * void_net )
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->GetNumberOfNodes();
}

int network_IsAcyclic( void * void_net )
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->IsAcyclic();
}

int network_ReadFile(void * void_net, char *thisFile)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->ReadFile(thisFile);
}

int network_WriteFile(void * void_net, char *thisFile)
{
	/*
	DSL_network.WriteFile( char *thisFile, int fileType=0)
	Writes the contents of the network to file
	*/
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->WriteFile(thisFile);
}

void network_SetDefaultBNAlgorithm(void * void_net, int theAlgorithm)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	net->SetDefaultBNAlgorithm(theAlgorithm);
}
void network_SetDefaultIDAlgorithm(void * void_net, int theAlgorithm)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	net->SetDefaultIDAlgorithm(theAlgorithm);
}
void network_SetDefaultHBNAlgorithm(void * void_net, int theAlgorithm)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	net->SetDefaultHBNAlgorithm(theAlgorithm);
}

int network_GetDefaultBNAlgorithm(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->GetDefaultBNAlgorithm();
}
int network_GetDefaultIDAlgorithm(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->GetDefaultIDAlgorithm();
}
int network_GetDefaultHBNAlgorithm(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->GetDefaultHBNAlgorithm();
}

int network_UpdateBeliefs(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->UpdateBeliefs();
}
int network_InvalidateAllBeliefs(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->InvalidateAllBeliefs();
}
int network_CallIDAlgorithm(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->CallIDAlgorithm();
}
int network_CallBNAlgorithm(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->CallBNAlgorithm();
}
int network_CallEqAlgorithm(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->CallEqAlgorithm();
}
int network_CallHBNAlgorithm(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->CallHBNAlgorithm();
}

int network_ClearAllEvidence(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->ClearAllEvidence();
}
int network_ClearAllDecision(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->ClearAllDecision();
}
int network_ClearAllPropagatedEvidence(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->ClearAllPropagatedEvidence();
}
int network_IsThereAnyEvidence(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->IsThereAnyEvidence();
}
int network_IsThereAnyDecision(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	return net->IsThereAnyDecision();
}

void * network_PartialOrdering(void * void_net)
{
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	DSL_intArray arr = net->PartialOrdering();
	DSL_intArray * arr2 = new DSL_intArray(arr);
	void * retval = arr2;
	return retval;
}